File: Problem/By section/Greedy/Problem 435/problem_435.java
import java.util.*;

class Solution {
    // Use Greedy Algothrim
    // Time: O(nlog(n)), Space: O(1)
    public int eraseOverlapIntervals(int[][] intervals) {
        if(intervals.length == 0) return 0;
        
        // Important!!!
        // How to define a comparator in order to do the sorting
        Arrays.sort(intervals, new Comparator<int[]>(){
            @Override
            public int compare(int[] arr1, int[] arr2){
                return arr1[1] - arr2[1];
            }
        });
        
        // Answer stands for the length of the longest legal interval
        // As long as the intervals array has length >= 1
        // we can at least have the head interval included in our final interval
        int ans = 1;
        int i = 0;
        int last_end = intervals[i][1];
        
        // iterate intervals array and include whatever the interval is leagal
        while(i + 1 < intervals.length ){
            if(intervals[++ i][0] >= last_end){
                ans ++;
                last_end = intervals[i][1];
            }
        }
        
        return intervals.length - ans;
    }
}

File: Problem/By section/Greedy/Problem 435/problem_435.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Time: O(nlogn), Space: O(1)
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        if(intervals.size() <= 1) return 0;
        // Important here, we only cares the end part 
        // counter example: {{1, 100}, {2, 3}, {4, 5}}
        auto compare = [](const vector<int> a, const vector<int> b){
            return a[1] < b[1];
        };
        sort(intervals.begin(), intervals.end(), compare);

        int result = 0;
        int prev = INT_MIN;
        for(auto interval : intervals) {
            if(interval[0] >= prev) {
                prev = interval[1];
            }else {
                result++;
            }
        }

        return result;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<vector<int>> input = {{1, 100}, {2, 3}, {4, 5}};
    cout << s.eraseOverlapIntervals(input) << endl;
}